utils_nlp/models/transformers/common.py

```diff
@@ -0,0 +1,277 @@
+        num_gpus = (
+            min(num_gpus, torch.cuda.device_count()) if num_gpus else torch.cuda.device_count()
+        )
+    else:
+        torch.cuda.set_device(local_rank)
+        device = torch.device("cuda", local_rank)
+        torch.distributed.init_process_group(backend="nccl")
+        num_gpus = 1
+
+    return device, num_gpus
+
+
+class Transformer:
+    def __init__(
+        self,
+        model_class,
+        model_name="bert-base-cased",
+        num_labels=2,
+        cache_dir=".",
+        load_model_from_dir=None,
+    ):
+        self.model_name = model_name
+        self.cache_dir = cache_dir
+        self.load_model_from_dir = load_model_from_dir
+        if load_model_from_dir is None:
+            self.model = model_class[model_name].from_pretrained(
+                model_name, cache_dir=cache_dir, num_labels=num_labels
+            )
+        else:
+            logger.info("Loading cached model from {}".format(load_model_from_dir))
+            self.model = model_class[model_name].from_pretrained(
+                load_model_from_dir, num_labels=num_labels
+            )
+
+    @property
+    def model_name(self):
+        return self._model_name
+
+    @model_name.setter
+    def model_name(self, value):
+        if value not in self.list_supported_models():
+            raise ValueError(
+                "Model name {0} is not supported by {1}. "
+                "Call '{2}.list_supported_models()' to get all supported model "
+                "names.".format(value, self.__class__.__name__, self.__class__.__name__)
+            )
+
+        self._model_name = value
+        self._model_type = value.split("-")[0]
+
+    @property
+    def model_type(self):
+        return self._model_type
+
+    @staticmethod
+    def set_seed(seed, cuda=True):
+        random.seed(seed)
+        np.random.seed(seed)
+        torch.manual_seed(seed)
+        if cuda and torch.cuda.is_available():
+            torch.cuda.manual_seed_all(seed)
+
+    def fine_tune(
+        self,
+        train_dataset,
+        get_inputs,
+        device,
+        max_steps=-1,
+        num_train_epochs=1,
+        max_grad_norm=1.0,
+        gradient_accumulation_steps=1,
+        per_gpu_train_batch_size=8,
+        n_gpu=1,
+        weight_decay=0.0,
+        learning_rate=5e-5,
+        adam_epsilon=1e-8,
+        warmup_steps=0,
+        fp16=False,
+        fp16_opt_level="O1",
+        local_rank=-1,
+        verbose=True,
+        seed=None,
+    ):
+        if seed is not None:
+            Transformer.set_seed(seed, n_gpu > 0)
+
+        train_batch_size = per_gpu_train_batch_size * max(1, n_gpu)
+        train_sampler = (
+            RandomSampler(train_dataset) if local_rank == -1 else DistributedSampler(train_dataset)
+        )
+        train_dataloader = DataLoader(
+            train_dataset, sampler=train_sampler, batch_size=train_batch_size
+        )
```

Comment:
Agree. Initially I thought this would save us from computing the batch size during the preprocessing step as we only need to pass it to fit(). I'll change this.